src/window/mod.rs

```diff
@@ -343,7 +371,21 @@ pub fn main_loop(
 
     let mut update_loop = UpdateLoop::new(cmd_line_settings.idle);
 
+    #[cfg(target_os = "macos")]
+    let mut menu_added = false;
     event_loop.run(move |e, window_target| {
+        #[cfg(target_os = "macos")]
+        if let (
+            Event::WindowEvent {
+                event: winit::event::WindowEvent::Focused(_),
+                ..
+            },
+            false,
+        ) = (&e, menu_added)
+        {
+            add_window_menu();
+            menu_added = true;
+        }
```

Comment:
the main loop function is getting long, I would like to see the same approach you did at `add_window_menu` making macOS specific changes external.

here for example, if we intend to add those kind of features on menu, as the project evolves would be hard to track and maintain.

I would do something like this to manage the menu state:

```rust

/// this should probably be moved to `/window/macos.rs` crate
#[cfg(target_os = "macos")]
struct MenuState {
    menu_added: bool,
}

impl MenuState {
    fn new() -> Self {
        Self { menu_added: false }
    }

    fn ensure_menu_added(&mut self, e: &Event<UserEvent>) {
        if let Event::WindowEvent {
            event: winit::event::WindowEvent::Focused(_),
            ..
        } = e
        {
            if !self.menu_added {
                add_window_menu();
                self.menu_added = true;
            }
        }
    }
}
```

and then use it in a simple way like:

```rust
#[cfg(target_os = "macos")]
let mut menu_state = MenuState::new();

event_loop.run(move |e, window_target| {
    #[cfg(target_os = "macos")]
    menu_state.ensure_menu_added(&e);

    if e == Event::LoopExiting {
        return;
    }

    if !RUNNING_TRACKER.is_running() {
        save_window_size(&window_wrapper);
        window_target.exit();
    } else {
        window_target.set_control_flow(update_loop.step(&mut window_wrapper, Ok(e)).unwrap());
    }
})
```

Reply:
Ok, did that